# Project
# EN.553.602 - Research and Design in Applied Mathematics: Data Mining

**Author**: Alex Teboul

**Professor**: Kearsley, Anthony José

**Data Source**: https://www.cdc.gov/brfss/annual_data/annual_2022.html

**Data SAS format** - 2022 BRFSS Data (SAS Transport Format) [ZIP – 64.3 MB]

**Code Book** - 2022 BRFSS Codebook CDC [ZIP – 3 MB]

## About this Project

**Objective:** The objective of this project was to develop predictive models for cognitive decline utilizing the 2022 BRFSS dataset. The workflow involved obtaining the 2022 BRFSS dataset, choosing features for examination in predictive models based on risk factors previously identified in studies on cognitive decline, conducting exploratory data analysis, testing various models, and presenting the findings. The methodologies investigated in this notebook include Random Forests, Gradient Boosting, AdaBoost, and Neural Networks.


1.   **Part 1:** Getting and Cleaning the Data
*   Get the BRFSS dataset
*   Select a Relevant Subset of Features
*   Cleaning the Data (Missing Values, Modifying Values, Make Feature Names More Readable, Save Finalized Dataset to CSV)
2.   **Part 2:** Model Building

Random Forests
*   Random Forest - w/ Feature Selection - Full Dataset
*   Random Forest - w/o Feature Selection - Full Dataset
*   Random Forest - w/ and w/o Feature Selection - 50-50 Balanced Dataset
*   Random Forest - w/ Feature Selection - 60-40 Balanced Dataset

AdaBoost, GradientBoost, and Neural Networks
*   AdaBoost, GradientBoost, and Neural Network - w/o Feature Selection - Full Dataset
*   AdaBoost, GradientBoost, and Neural Network - w/ Feature Selection - Full Dataset
*   AdaBoost, GradientBoost, and Neural Network - w/ Feature Selection - 50-50 Dataset
*   AdaBoost, GradientBoost, and Neural Network - w/ Feature Selection - 60-40 Dataset

Support Vector Machines: *Too Slow - Never Finishes
*   RBF-SVM - w/ Feature Selection 50-50 Dataset




# Part 1: Getting and Cleaning the Data

In [2]:
#imports
import pandas as pd
import random
import numpy as np
random.seed(1)

### Get the BRFSS dataset

In [6]:
df=pd.read_sas('/content/drive/MyDrive/Colab Notebooks/Data Mining/Data/LLCP2022.XPT ', format='xport')
df.head()

FileNotFoundError: [Errno 2] No such file or directory: '/content/drive/MyDrive/Colab Notebooks/Data Mining/Data/LLCP2022.XPT '

In [ ]:
#check that all the data loaded in
df.shape

(445132, 328)

In [ ]:
#check that the data loaded in is in the correct format
pd.set_option('display.max_columns', 500)
df.head()

_STATE  FMONTH        IDATE IMONTH   IDAY    IYEAR  DISPCODE  \
0     1.0     1.0  b'02032022'  b'02'  b'03'  b'2022'    1100.0   
1     1.0     1.0  b'02042022'  b'02'  b'04'  b'2022'    1100.0   
2     1.0     1.0  b'02022022'  b'02'  b'02'  b'2022'    1100.0   
3     1.0     1.0  b'02032022'  b'02'  b'03'  b'2022'    1100.0   
4     1.0     1.0  b'02022022'  b'02'  b'02'  b'2022'    1100.0   

           SEQNO          _PSU  CTELENM1  PVTRESD1  COLGHOUS  STATERE1  \
0  b'2022000001'  2.022000e+09       1.0       1.0       NaN       1.0   
1  b'2022000002'  2.022000e+09       1.0       1.0       NaN       1.0   
2  b'2022000003'  2.022000e+09       1.0       1.0       NaN       1.0   
3  b'2022000004'  2.022000e+09       1.0       1.0       NaN       1.0   
4  b'2022000005'  2.022000e+09       1.0       1.0       NaN       1.0   

   CELPHON1  LADULT1  COLGSEX1  NUMADULT  LANDSEX1  NUMMEN  NUMWOMEN  \
0       2.0      1.0       NaN       2.0       NaN     1.0       1.0   
1       2.0      1.0       NaN       2.0       NaN     1.0       1.0   
2       2.0      1.0       NaN       1.0       2.0     NaN       NaN   
3       2.0      1.0       NaN       3.0       NaN     2.0       1.0   
4       2.0      1.0       NaN       2.0       NaN     1.0       1.0   

   RESPSLCT  SAFETIME  CTELNUM1  CELLFON5  CADULT1  CELLSEX1  PVTRESD3  \
0       2.0       NaN       NaN       NaN      NaN       NaN       NaN   
1       2.0       NaN       NaN       NaN      NaN       NaN       NaN   
2       NaN       NaN       NaN       NaN      NaN       NaN       NaN   
3       2.0       NaN       NaN       NaN      NaN       NaN       NaN   
4       2.0       NaN       NaN       NaN      NaN       NaN       NaN   

   CCLGHOUS  CSTATE1  LANDLINE  HHADULT  SEXVAR  GENHLTH  PHYSHLTH  MENTHLTH  \
0       NaN      NaN       NaN      NaN     2.0      2.0      88.0      88.0   
1       NaN      NaN       NaN      NaN     2.0      1.0      88.0      88.0   
2       NaN      NaN       NaN      NaN     2.0      2.0       2.0       3.0   
3       NaN      NaN       NaN      NaN     2.0      1.0      88.0      88.0   
4       NaN      NaN       NaN      NaN     2.0      4.0       2.0      88.0   

   POORHLTH  PRIMINSR  PERSDOC3  MEDCOST1  CHECKUP1  EXERANY2  SLEPTIM1  \
0       NaN      99.0       1.0       2.0       1.0       2.0       8.0   
1       NaN       3.0       2.0       2.0       8.0       2.0       6.0   
2       2.0       1.0       1.0       2.0       1.0       1.0       5.0   
3       NaN      99.0       1.0       2.0       1.0       1.0       7.0   
4      88.0       7.0       2.0       2.0       1.0       1.0       9.0   

   LASTDEN4  RMVTETH4  CVDINFR4  CVDCRHD4  CVDSTRK3  ASTHMA3  ASTHNOW  \
0       NaN       NaN       2.0       2.0       2.0      2.0      NaN   
1       NaN       NaN       2.0       2.0       2.0      2.0      NaN   
2       NaN       NaN       2.0       2.0       2.0      2.0      NaN   
3       NaN       NaN       2.0       2.0       2.0      1.0      1.0   
4       NaN       NaN       2.0       2.0       2.0      2.0      NaN   

   CHCSCNC1  CHCOCNC1  CHCCOPD3  ADDEPEV3  CHCKDNY2  HAVARTH4  DIABETE4  \
0       2.0       2.0       2.0       2.0       2.0       2.0       1.0   
1       1.0       1.0       2.0       2.0       2.0       2.0       3.0   
2       1.0       2.0       2.0       2.0       2.0       2.0       3.0   
3       2.0       2.0       2.0       2.0       2.0       1.0       3.0   
4       2.0       2.0       2.0       2.0       2.0       2.0       3.0   

   DIABAGE4  MARITAL  EDUCA  RENTHOM1  NUMHHOL4  NUMPHON4  CPDEMO1C  VETERAN3  \
0      80.0      1.0    6.0       1.0       1.0       1.0       2.0       2.0   
1       NaN      3.0    4.0       1.0       1.0       2.0       1.0       2.0   
2       NaN      1.0    6.0       1.0       2.0       NaN       1.0       2.0   
3       NaN      1.0    4.0       1.0       2.0       NaN       1.0       2.0   
4       NaN      1.0    5.0       1.0       2.0   

### Select Relevant Subset of Features

The dataset originally has 328 features (columns), but based on Cognitive decline research regarding factors influencing Cognitive decline, only select features are included in this analysis.



#### Important Risk Factors
Research in the field has identified the following as **important risk factors** for Cognitive decline:

*   blood pressure (high)
*   cholesterol (high)
*   smoking
*   diabetes
*   obesity
*   age
*   sex
*   race
*   diet
*   exercise
*   alcohol consumption
*   BMI
*   Household Income
*   Marital Status
*   Sleep
*   Time since last checkup
*   Education
*   Health care coverage
*   Mental Health



#### Selected Subset of Features from BRFSS
Given these risk factors, I tried to select features (columns/questions) in the BRFSS related to these risk factors. To help understand what the columns mean, I consult the BRFSS Codebook to see the questions and information about the questions. I try to match the variable names in the codebook to the variable names in the dataset.

The **selected features** from the BRFSS dataset are:

**Response Variable / Dependent Variable:**
*   Label: Have you experienced confusion or memory loss that is happening more often or is getting worse?
*   SAS Variable Name: CIMEMLOS

**Independent Variables:**


#### Get Subset of Features

In [ ]:
# select specific columns
brfss_df_selected = df[['CIMEMLOS','_BMI5','SMOKE100',
                        'CVDSTRK3', '_SEX',
                        '_TOTINDA',
                        'GENHLTH', 'MENTHLTH', 'PHYSHLTH', 'DIFFWALK',
                        '_AGEG5YR', 'EDUCA','ACEDEPRS','ADDEPEV3' ]]


In [ ]:
brfss_df_selected.shape

(445132, 14)

In [ ]:
brfss_df_selected.head()

CIMEMLOS   _BMI5  SMOKE100  CVDSTRK3  _SEX  _TOTINDA  GENHLTH  MENTHLTH  \
0       NaN     NaN       2.0       2.0   2.0       2.0      2.0      88.0   
1       NaN  2657.0       2.0       2.0   2.0       2.0      1.0      88.0   
2       NaN  2561.0       2.0       2.0   2.0       1.0      2.0       3.0   
3       NaN  2330.0       1.0       2.0   2.0       1.0      1.0      88.0   
4       NaN  2177.0       2.0       2.0   2.0       1.0      4.0      88.0   

   PHYSHLTH  DIFFWALK  _AGEG5YR  EDUCA  ACEDEPRS  ADDEPEV3  
0      88.0       2.0      13.0    6.0       NaN       2.0  
1      88.0       2.0      13.0    4.0       NaN       2.0  
2       2.0       2.0       8.0    6.0       NaN       2.0  
3      88.0       2.0      14.0    4.0       NaN       2.0  
4       2.0       2.0       5.0    5.0       NaN       2.0

### Cleaning the Data

#### Missing Values

In [ ]:
#Drop Missing Values - knocks 100,000 rows out right away
brfss_df_selected = brfss_df_selected.dropna()
brfss_df_selected.shape

(17851, 14)

In [ ]:
brfss_df_selected.head()

CIMEMLOS   _BMI5  SMOKE100  CVDSTRK3  _SEX  _TOTINDA  GENHLTH  \
63190       1.0  1953.0       2.0       2.0   1.0       1.0      2.0   
63191       2.0  3109.0       1.0       2.0   2.0       1.0      2.0   
63193       2.0  2439.0       1.0       2.0   1.0       1.0      1.0   
63194       2.0  3228.0       2.0       2.0   1.0       1.0      4.0   
63195       2.0  2918.0       1.0       2.0   2.0       1.0      2.0   

       MENTHLTH  PHYSHLTH  DIFFWALK  _AGEG5YR  EDUCA  ACEDEPRS  ADDEPEV3  
63190      88.0      88.0       2.0      12.0    6.0       2.0       2.0  
63191      88.0      88.0       2.0       9.0    6.0       2.0       2.0  
63193      88.0      88.0       7.0      13.0    6.0       2.0       2.0  
63194       3.0       1.0       1.0      10.0    6.0       2.0       2.0  
63195      88.0      88.0       2.0      10.0    5.0       1.0       2.0

#### Modifying Values

In [ ]:
# CIMEMLOS
#Change 2 to 0 because this means did not have MI or CHD
brfss_df_selected['CIMEMLOS'] = brfss_df_selected['CIMEMLOS'].replace({2: 0})
brfss_df_selected.CIMEMLOS.unique()

array([1., 0., 7., 9.])

In [ ]:
# Remove rows where 'CIMEMLOS' contains 7 or 9
brfss_df_selected = brfss_df_selected[~brfss_df_selected['CIMEMLOS'].isin([7, 9])]
brfss_df_selected.CIMEMLOS.unique()

array([1., 0.])

In [ ]:
# ACEDEPRS -  Live With Anyone Depressed, Mentally Ill, Or Suicidal?
#Change 2 to 0 for no
brfss_df_selected['ACEDEPRS'] = brfss_df_selected['ACEDEPRS'].replace({2: 0})
brfss_df_selected['ACEDEPRS'] = brfss_df_selected['ACEDEPRS'].fillna(7)
brfss_df_selected = brfss_df_selected[brfss_df_selected.ACEDEPRS != 7]
brfss_df_selected = brfss_df_selected[brfss_df_selected.ACEDEPRS != 9]
brfss_df_selected.ACEDEPRS.unique()

array([0., 1.])

In [ ]:
# ADDEPEV3 -   you had a depressive disorde
#Change 2 to 0 for no
brfss_df_selected['ADDEPEV3'] = brfss_df_selected['ADDEPEV3'].replace({2: 0})
brfss_df_selected['ADDEPEV3'] = brfss_df_selected['ADDEPEV3'].fillna(7)
brfss_df_selected = brfss_df_selected[brfss_df_selected.ADDEPEV3 != 7]
brfss_df_selected = brfss_df_selected[brfss_df_selected.ADDEPEV3 != 9]
brfss_df_selected.ADDEPEV3.unique()

array([0., 1.])

In [ ]:
# _SEX
#Change 2 to 0 for female
brfss_df_selected['_SEX'] = brfss_df_selected['_SEX'].replace({2: 0})
brfss_df_selected._SEX.unique()

array([1., 0.])

In [ ]:
#4 _BMI5 (no changes, just note that these are BMI * 100. So for example a BMI of 4018 is really 40.18)
brfss_df_selected['_BMI5'] = brfss_df_selected['_BMI5'].div(100).round(0)
brfss_df_selected._BMI5.unique()

array([20., 31., 24., 32., 29., 52., 23., 36., 21., 22., 46., 35., 19.,
       30., 43., 25., 27., 28., 34., 26., 33., 42., 37., 38., 47., 45.,
       44., 39., 18., 17., 49., 41., 16., 15., 40., 51., 48., 66., 54.,
       53., 14., 96., 61., 56., 13., 57., 55., 50., 80., 63., 65., 62.,
       73., 83., 59., 60., 67., 94., 72., 84., 12., 70., 78.])

In [ ]:
#5 SMOKE100
# Change 2 to 0 because it is No
# Remove all 7 (dont knows)
# Remove all 9 (refused)
brfss_df_selected['SMOKE100'] = brfss_df_selected['SMOKE100'].replace({2:0})
brfss_df_selected = brfss_df_selected[brfss_df_selected.SMOKE100 != 7]
brfss_df_selected = brfss_df_selected[brfss_df_selected.SMOKE100 != 9]
brfss_df_selected.SMOKE100.unique()

array([0., 1.])

In [ ]:
#6 CVDSTRK3
# Change 2 to 0 because it is No
# Remove all 7 (dont knows)
# Remove all 9 (refused)
brfss_df_selected['CVDSTRK3'] = brfss_df_selected['CVDSTRK3'].replace({2:0})
brfss_df_selected = brfss_df_selected[brfss_df_selected.CVDSTRK3 != 7]
brfss_df_selected = brfss_df_selected[brfss_df_selected.CVDSTRK3 != 9]
brfss_df_selected.CVDSTRK3.unique()

array([0., 1.])

In [ ]:
#8 _TOTINDA
# 1 for physical activity
# change 2 to 0 for no physical activity
# Remove all 9 (don't know/refused)
brfss_df_selected['_TOTINDA'] = brfss_df_selected['_TOTINDA'].replace({2:0})
brfss_df_selected = brfss_df_selected[brfss_df_selected._TOTINDA != 9]
brfss_df_selected._TOTINDA.unique()

array([1., 0.])

In [ ]:
#14 GENHLTH
# This is an ordinal variable that I want to keep (1 is Excellent -> 5 is Poor)
# Remove 7 and 9 for don't know and refused
brfss_df_selected = brfss_df_selected[brfss_df_selected.GENHLTH != 7]
brfss_df_selected = brfss_df_selected[brfss_df_selected.GENHLTH != 9]
brfss_df_selected.GENHLTH.unique()

array([2., 1., 4., 3., 5.])

In [ ]:
#15 MENTHLTH
# already in days so keep that, scale will be 0-30
# change 88 to 0 because it means none (no bad mental health days)
# remove 77 and 99 for don't know not sure and refused
brfss_df_selected['MENTHLTH'] = brfss_df_selected['MENTHLTH'].replace({88:0})
brfss_df_selected = brfss_df_selected[brfss_df_selected.MENTHLTH != 77]
brfss_df_selected = brfss_df_selected[brfss_df_selected.MENTHLTH != 99]
brfss_df_selected.MENTHLTH.unique()

array([ 0.,  3.,  2.,  1., 30.,  5., 10., 14., 12., 15.,  4.,  8.,  6.,
        7., 20., 29.,  9., 25., 21., 26., 18., 28., 23., 16., 11., 27.,
       13., 17., 22.])

In [ ]:
#16 PHYSHLTH
# already in days so keep that, scale will be 0-30
# change 88 to 0 because it means none (no bad mental health days)
# remove 77 and 99 for don't know not sure and refused
brfss_df_selected['PHYSHLTH'] = brfss_df_selected['PHYSHLTH'].replace({88:0})
brfss_df_selected = brfss_df_selected[brfss_df_selected.PHYSHLTH != 77]
brfss_df_selected = brfss_df_selected[brfss_df_selected.PHYSHLTH != 99]
brfss_df_selected.PHYSHLTH.unique()

array([ 0.,  1.,  3., 10., 20., 30., 14., 15.,  2.,  5.,  4.,  7., 13.,
       29., 28.,  8., 17., 21., 19.,  6., 25., 26., 18., 12., 16.,  9.,
       11., 22., 23., 24., 27.])

In [ ]:
#17 DIFFWALK
# change 2 to 0 for no. 1 is already yes
# remove 7 and 9 for don't know not sure and refused
brfss_df_selected['DIFFWALK'] = brfss_df_selected['DIFFWALK'].replace({2:0})
brfss_df_selected = brfss_df_selected[brfss_df_selected.DIFFWALK != 7]
brfss_df_selected = brfss_df_selected[brfss_df_selected.DIFFWALK != 9]
brfss_df_selected.DIFFWALK.unique()

array([0., 1.])

In [ ]:
#19 _AGEG5YR
# already ordinal. 1 is 18-24 all the way up to 13 wis 80 and older. 5 year increments.
# remove 14 because it is don't know or missing
brfss_df_selected = brfss_df_selected[brfss_df_selected._AGEG5YR != 14]
brfss_df_selected._AGEG5YR.unique()

array([12.,  9., 10.,  6., 11.,  7., 13.,  8.])

In [ ]:
#20 EDUCA
# This is already an ordinal variable with 1 being never attended school or kindergarten only up to 6 being college 4 years or more
# Scale here is 1-6
# Remove 9 for refused:
brfss_df_selected = brfss_df_selected[brfss_df_selected.EDUCA != 9]
brfss_df_selected.EDUCA.unique()

array([6., 5., 4., 3., 2., 1.])

In [ ]:
#Check the shape of the dataset now: We have 253,680 cleaned rows and 22 columns (1 of which is our dependent variable)
brfss_df_selected.shape

(16031, 14)

In [ ]:
 #Check Class Sizes
 brfss_df_selected.groupby(['CIMEMLOS']).size()

CIMEMLOS
0.0    14221
1.0     1810
dtype: int64

#### Make Feature Names More Readable

In [ ]:
 #Rename the columns to make them more readable
 brfss = brfss_df_selected.rename(columns = {'_BMI5':'BMI',
                                         'SMOKE100':'Smoker',
                                         'CVDSTRK3':'Stroke',
                                         '_TOTINDA':'PhysActivity',
                                         'GENHLTH':'GenHlth',
                                             'MENTHLTH':'MentHlth',
                                             'PHYSHLTH':'PhysHlth',
                                             'DIFFWALK':'DiffWalk',
                                          '_AGEG5YR':'Age',
                                             'EDUCA':'Education',
                                             'CIMEMLOS':'CognitiveDecline',
                                            'ACEDEPRS':'liveWithDepressed',
                                            'ADDEPEV3':'toldDepression',
                                            '_SEX':'Sex'})

In [ ]:
#See the cleaned dataset
brfss.head(10)

CognitiveDecline   BMI  Smoker  Stroke  Sex  PhysActivity  GenHlth  \
63190               1.0  20.0     0.0     0.0  1.0           1.0      2.0   
63191               0.0  31.0     1.0     0.0  0.0           1.0      2.0   
63194               0.0  32.0     0.0     0.0  1.0           1.0      4.0   
63195               0.0  29.0     1.0     0.0  0.0           1.0      2.0   
63196               0.0  24.0     0.0     0.0  0.0           1.0      3.0   
63197               0.0  52.0     0.0     0.0  0.0           1.0      2.0   
63198               0.0  23.0     0.0     0.0  0.0           1.0      2.0   
63199               0.0  36.0     0.0     0.0  1.0           1.0      2.0   
63200               0.0  31.0     1.0     0.0  1.0           0.0      4.0   
63201               0.0  21.0     1.0     0.0  0.0           0.0      4.0   

       MentHlth  PhysHlth  DiffWalk   Age  Education  liveWithDepressed  \
63190       0.0       0.0       0.0  12.0        6.0                0.0   
63191       0.0       0.0       0.0   9.0        6.0                0.0   
63194       3.0       1.0       1.0  10.0        6.0                0.0   
63195       0.0       0.0       0.0  10.0        5.0                1.0   
63196       2.0       3.0       0.0   6.0        4.0                0.0   
63197       0.0       0.0       1.0   9.0        4.0                0.0   
63198       0.0       1.0       0.0  10.0        6.0                0.0   
63199       3.0       0.0       0.0  11.0        5.0                1.0   
63200       1.0      10.0       0.0  11.0        6.0                0.0   
63201      30.0      20.0       1.0  10.0        4.0                0.0   

       toldDepression  
63190             0.0  
63191             0.0  
63194             0.0  
63195             0.0  
63196             0.0  
63197             0.0  
63198             0.0  
63199             0.0  
63200             0.0  
63201             1.0

In [ ]:

correlation_matrix = brfss.corr()

# Extract the correlation of cognitive decline with other variables
cognitive_decline_corr = correlation_matrix['CognitiveDecline'].drop('CognitiveDecline')
print("Correlation of Cognitive Decline with other variables:")
print(cognitive_decline_corr)


Correlation of Cognitive Decline with other variables:
BMI                  0.040355
Smoker               0.053978
Stroke               0.110135
Sex                 -0.007527
PhysActivity        -0.108755
GenHlth              0.240656
MentHlth             0.234320
PhysHlth             0.227483
DiffWalk             0.223705
Age                  0.041151
Education           -0.071946
liveWithDepressed    0.120254
toldDepression       0.213583
Name: CognitiveDecline, dtype: float64


In [ ]:
# Drop the specified columns
brfss.drop(['Sex', 'Education', 'PhysActivity'], axis=1, inplace=True)


In [ ]:
correlation_matrix = brfss.corr()

# Extract the correlation of cognitive decline with other variables
cognitive_decline_corr = correlation_matrix['CognitiveDecline'].drop('CognitiveDecline')
print("Correlation of Cognitive Decline with other variables:")
print(cognitive_decline_corr)

Correlation of Cognitive Decline with other variables:
BMI                  0.040355
Smoker               0.053978
Stroke               0.110135
GenHlth              0.240656
MentHlth             0.234320
PhysHlth             0.227483
DiffWalk             0.223705
Age                  0.041151
liveWithDepressed    0.120254
toldDepression       0.213583
Name: CognitiveDecline, dtype: float64


In [ ]:
#Double check shape of the dataset (rows and columns)
brfss.shape

(16031, 11)

In [ ]:
 #Check how many respondents have had heart disease or a heart attack. Note the class imbalance!
 brfss.groupby(['CognitiveDecline']).size()

CognitiveDecline
0.0    14221
1.0     1810
dtype: int64

#### Save Finalized Dataset to CSV

In [ ]:
#************************************************************************************************
brfss.to_csv('brfss_cleaned.csv', sep=",", index=False)
#************************************************************************************************

#### Get a BALANCED 50-50 Dataset Randomly Selected
*  The brfss dataset is clearly imbalanced. When training my models, I get about 90% accuracy on many models with AUC between 70 and 80. This may be caused by the models are learning the distribution in the data.
*  To check these concerns, I will create a second dataset with a 50-50 balance for the HeartDiseaseorAttack response variable - just to compare performance.
*  To do this, I will take a random sample of 49786 instances of the 0 (or No Cognitive Decline) and all of the 6003 instances of the 1 (or Yes).
* The if the new dataset performs comparably, then I can rest assured that it
* With roughly 48,000 datapoints, I hope that this is sufficient to train the model and that the random selection will not greatly change the results. I have the random seed set to 1.

In [ ]:
#Separate the 0 and 1
#Get the 1s
is1 = brfss['CognitiveDecline'] == 1
brfss_5050_1 = brfss[is1]
# Check the type of brfss_5050
print(type(brfss_5050_1))

#Get the 0s
is0 = brfss['CognitiveDecline'] == 0
brfss_5050_0 = brfss[is0]
print(type(brfss_5050_0))

#Select the 6003 random cases for 0
brfss_5050_0_rand1 = brfss_5050_0.take(np.random.permutation(len(brfss_5050_0))[:6003])
print(type(brfss_5050_0_rand1))

brfss_5050 = pd.concat([brfss_5050_0_rand1, brfss_5050_1], ignore_index=True)

<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.frame.DataFrame'>


In [ ]:
#Check that it worked. Now we have a dataset of 47,786 rows that is equally balanced with 50% 1 and 50% 0 for the target variable HeartDiseaseorAttack
brfss_5050

CognitiveDecline   BMI  Smoker  Stroke  GenHlth  MentHlth  PhysHlth  \
0                  0.0  27.0     1.0     1.0      4.0       0.0      30.0   
1                  0.0  37.0     0.0     0.0      4.0       0.0      14.0   
2                  0.0  31.0     1.0     0.0      3.0       0.0       5.0   
3                  0.0  24.0     1.0     0.0      2.0       0.0       0.0   
4                  0.0  20.0     0.0     0.0      2.0       0.0       0.0   
...                ...   ...     ...     ...      ...       ...       ...   
7808               1.0  28.0     1.0     0.0      4.0      20.0      30.0   
7809               1.0  27.0     1.0     0.0      3.0      15.0       0.0   
7810               1.0  27.0     0.0     0.0      5.0       7.0       7.0   
7811               1.0  26.0     0.0     1.0      5.0      15.0      30.0   
7812               1.0  41.0     0.0     0.0      1.0       0.0       0.0   

      DiffWalk   Age  liveWithDepressed  toldDepression  
0          1.0   9.0                0.0             0.0  
1          0.0   9.0                0.0             0.0  
2          0.0  12.0                0.0             0.0  
3          0.0  13.0                0.0             0.0  
4          0.0  12.0                0.0             0.0  
...        ...   ...                ...             ...  
7808       0.0  10.0                0.0             1.0  
7809       1.0   9.0                0.0             0.0  
7810       1.0   8.0                0.0             1.0  
7811       1.0   6.0                0.0             1.0  
7812       0.0   7.0                1.0             1.0  

[7813 rows x 11 columns]

In [ ]:
#See the classes are perfectly balanced now
brfss_5050.groupby(['CognitiveDecline']).size()

CognitiveDecline
0.0    6003
1.0    1810
dtype: int64

In [ ]:
#Save the 50-50 balanced dataset to csv

#************************************************************************************************
brfss_5050.to_csv('brfss_5050_cleaned.csv', sep=",", index=False)
#************************************************************************************************

#### Also Get a 60-40 Dataset Randomly Selected

In [ ]:
#Also make a 60-40 dataset
brfss_6040_0_rand1 = brfss_5050_0.take(np.random.permutation(len(brfss_5050_0))[:12006])


brfss_6040 =  pd.concat([brfss_6040_0_rand1,brfss_5050_1], ignore_index = True)
#Save the 6040 balanced dataset to csv
#************************************************************************************************
brfss_6040.to_csv('brfss_6040_cleaned.csv', sep=",", index=False)
#************************************************************************************************
brfss_6040

CognitiveDecline   BMI  Smoker  Stroke  GenHlth  MentHlth  PhysHlth  \
0                   0.0  25.0     1.0     0.0      4.0      15.0      30.0   
1                   0.0  33.0     0.0     0.0      1.0       0.0       0.0   
2                   0.0  33.0     1.0     0.0      4.0       0.0       4.0   
3                   0.0  28.0     1.0     0.0      4.0      10.0       4.0   
4                   0.0  28.0     1.0     0.0      3.0       0.0       0.0   
...                 ...   ...     ...     ...      ...       ...       ...   
13811               1.0  28.0     1.0     0.0      4.0      20.0      30.0   
13812               1.0  27.0     1.0     0.0      3.0      15.0       0.0   
13813               1.0  27.0     0.0     0.0      5.0       7.0       7.0   
13814               1.0  26.0     0.0     1.0      5.0      15.0      30.0   
13815               1.0  41.0     0.0     0.0      1.0       0.0       0.0   

       DiffWalk   Age  liveWithDepressed  toldDepression  
0           1.0   9.0                0.0             0.0  
1           0.0   8.0                0.0             0.0  
2           0.0  11.0                0.0             0.0  
3           1.0   9.0                0.0             0.0  
4           0.0   7.0                0.0             0.0  
...         ...   ...                ...             ...  
13811       0.0  10.0                0.0             1.0  
13812       1.0   9.0                0.0             0.0  
13813       1.0   8.0                0.0             1.0  
13814       1.0   6.0                0.0             1.0  
13815       0.0   7.0                1.0             1.0  

[13816 rows x 11 columns]

# Part 2: Model Building

## Random Forests

In [ ]:
df_full = pd.read_csv('brfss_cleaned.csv')
df_full.head()

CognitiveDecline   BMI  Smoker  Stroke  GenHlth  MentHlth  PhysHlth  \
0               1.0  20.0     0.0     0.0      2.0       0.0       0.0   
1               0.0  31.0     1.0     0.0      2.0       0.0       0.0   
2               0.0  32.0     0.0     0.0      4.0       3.0       1.0   
3               0.0  29.0     1.0     0.0      2.0       0.0       0.0   
4               0.0  24.0     0.0     0.0      3.0       2.0       3.0   

   DiffWalk   Age  liveWithDepressed  toldDepression  
0       0.0  12.0                0.0             0.0  
1       0.0   9.0                0.0             0.0  
2       1.0  10.0                0.0             0.0  
3       0.0  10.0                1.0             0.0  
4       0.0   6.0                0.0             0.0

In [ ]:
df_5050 = pd.read_csv('brfss_5050_cleaned.csv')
df_5050.head()

CognitiveDecline   BMI  Smoker  Stroke  GenHlth  MentHlth  PhysHlth  \
0               0.0  27.0     1.0     1.0      4.0       0.0      30.0   
1               0.0  37.0     0.0     0.0      4.0       0.0      14.0   
2               0.0  31.0     1.0     0.0      3.0       0.0       5.0   
3               0.0  24.0     1.0     0.0      2.0       0.0       0.0   
4               0.0  20.0     0.0     0.0      2.0       0.0       0.0   

   DiffWalk   Age  liveWithDepressed  toldDepression  
0       1.0   9.0                0.0             0.0  
1       0.0   9.0                0.0             0.0  
2       0.0  12.0                0.0             0.0  
3       0.0  13.0                0.0             0.0  
4       0.0  12.0                0.0             0.0

In [ ]:
df_6040 = pd.read_csv('brfss_6040_cleaned.csv')
df_6040.head()

CognitiveDecline   BMI  Smoker  Stroke  GenHlth  MentHlth  PhysHlth  \
0               0.0  25.0     1.0     0.0      4.0      15.0      30.0   
1               0.0  33.0     0.0     0.0      1.0       0.0       0.0   
2               0.0  33.0     1.0     0.0      4.0       0.0       4.0   
3               0.0  28.0     1.0     0.0      4.0      10.0       4.0   
4               0.0  28.0     1.0     0.0      3.0       0.0       0.0   

   DiffWalk   Age  liveWithDepressed  toldDepression  
0       1.0   9.0                0.0             0.0  
1       0.0   8.0                0.0             0.0  
2       0.0  11.0                0.0             0.0  
3       1.0   9.0                0.0             0.0  
4       0.0   7.0                0.0             0.0

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, roc_auc_score, confusion_matrix, precision_score

# Load dataframes
dfs = {
    'df_full': pd.read_csv('brfss_cleaned.csv'),
    'df_6040': pd.read_csv('brfss_6040_cleaned.csv'),
    'df_5050': pd.read_csv('brfss_5050_cleaned.csv')
}

# Loop over each dataframe
for df_name, df in dfs.items():
    print(f"\nRunning RandomForestClassifier on {df_name}...")

    # Separate features (X) and target variable (y)
    X = df.drop(columns=["CognitiveDecline"])
    y = df["CognitiveDecline"]

    # Split the data into training and testing sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Initialize the random forest classifier
    rf_classifier = RandomForestClassifier(n_estimators=100, random_state=42)

    # Train the classifier on the training data
    rf_classifier.fit(X_train, y_train)

    # Make predictions on the testing data
    y_pred = rf_classifier.predict(X_test)

    # Calculate the accuracy of the model
    accuracy = accuracy_score(y_test, y_pred)
    print(f"Accuracy of {df_name}: {accuracy:.4f}")

    # Calculate the AUC score
    auc_score = roc_auc_score(y_test, y_pred)
    print(f"AUC of {df_name}: {auc_score:.4f}")

    # Calculate the confusion matrix
    conf_matrix = confusion_matrix(y_test, y_pred)
    print(f"Confusion Matrix of {df_name}:\n{conf_matrix}")

    # Calculate the precision score
    precision = precision_score(y_test, y_pred)
    print(f"Precision of {df_name}: {precision:.4f}")



Running RandomForestClassifier on df_full...
Accuracy of df_full: 0.8803
AUC of df_full: 0.5504
Confusion Matrix of df_full:
[[2778   56]
 [ 328   45]]
Precision of df_full: 0.4455

Running RandomForestClassifier on df_6040...
Accuracy of df_6040: 0.8614
AUC of df_6040: 0.5669
Confusion Matrix of df_6040:
[[2320   69]
 [ 314   61]]
Precision of df_6040: 0.4692

Running RandomForestClassifier on df_5050...
Accuracy of df_5050: 0.7646
AUC of df_5050: 0.6080
Confusion Matrix of df_5050:
[[1080  113]
 [ 255  115]]
Precision of df_5050: 0.5044


In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import accuracy_score

# Load dataframes
dfs = {
    'df_full': pd.read_csv('brfss_cleaned.csv'),
    'df_6040': pd.read_csv('brfss_6040_cleaned.csv'),
    'df_5050': pd.read_csv('brfss_5050_cleaned.csv')
}

# Loop over each dataframe
for df_name, df in dfs.items():
    print(f"Running GradientBoostingClassifier on {df_name}...")

    # Separate features (X) and target variable (y)
    X = df.drop(columns=["CognitiveDecline"])
    y = df["CognitiveDecline"]

    # Split the data into training and testing sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Initialize the gradient boosting classifier
    gb_classifier = GradientBoostingClassifier(n_estimators=100, random_state=42)

    # Train the classifier on the training data
    gb_classifier.fit(X_train, y_train)

    # Make predictions on the testing data
    y_pred = gb_classifier.predict(X_test)

    # Calculate the accuracy of the model
    accuracy = accuracy_score(y_test, y_pred)
    print(f"Accuracy of {df_name}: {accuracy}\n")


Running GradientBoostingClassifier on df_full...
Accuracy of df_full: 0.888992828188338

Running GradientBoostingClassifier on df_6040...
Accuracy of df_6040: 0.8726483357452967

Running GradientBoostingClassifier on df_5050...
Accuracy of df_5050: 0.7779910428662828



In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
import tensorflow as tf
from sklearn.metrics import accuracy_score

# Load dataframes
dfs = {
    'df_full': pd.read_csv('brfss_cleaned.csv'),
    'df_6040': pd.read_csv('brfss_6040_cleaned.csv'),
    'df_5050': pd.read_csv('brfss_5050_cleaned.csv')
}

# Define a simple neural network model
def create_model(input_shape):
    model = tf.keras.Sequential([
        tf.keras.layers.Dense(64, activation='relu', input_shape=input_shape),
        tf.keras.layers.Dense(32, activation='relu'),
        tf.keras.layers.Dense(1, activation='sigmoid')
    ])
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    return model

# Loop over each dataframe
for df_name, df in dfs.items():
    print(f"Running Neural Network on {df_name}...")

    # Separate features (X) and target variable (y)
    X = df.drop(columns=["CognitiveDecline"])
    y = df["CognitiveDecline"]

    # Split the data into training and testing sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Create and compile the neural network model
    model = create_model(input_shape=(X_train.shape[1],))

    # Train the model on the training data
    model.fit(X_train, y_train, epochs=10, batch_size=32, verbose=0)

    # Evaluate the model on the testing data
    _, accuracy = model.evaluate(X_test, y_test, verbose=0)
    print(f"Accuracy of {df_name}: {accuracy}\n")


Running Neural Network on df_full...
Accuracy of df_full: 0.8861864805221558

Running Neural Network on df_6040...
Accuracy of df_6040: 0.8737336993217468

Running Neural Network on df_5050...
Accuracy of df_5050: 0.782469630241394

